### DISCUSSION TITLE
Hello, a small question about the event queue design

### DISCUSSION THREAD

**Discussion by cmzxo** [gbeced/basana#45] [General]
Hello, I'm new to basana (masterpiece code really!), and very interested in its event queue design.

I had some trading scripts written by myself, which also have a Producer (Broker) to receive streaming data from broker then call all subscribed callbacks (by Strategies) directly. So the control flow (callback chain) is very simple for me.

Then I found basana, it uses a very cool event queue design, decouple every component by just using event! But I have a small question about its callbacks execution order (maybe I need to read its source code).

See, event dispatcher pop event A, which has two callbacks subscribed callback1 and callback2. The dispatcher would execute callback1 first, what if it push another event B (subscribed by callback3). My question is the execution orders of callback2 and callback3. Will the callback3 be executed after callback2?

If using simple callback chain, callback3 will be executed before callback2, like:
```
A -> callback1 (B) -> callback3
   -> callback2
```
Maybe I missed something, thanks!

**Comment by gbeced** [comment]
When events are pushed they get pushed into a queue, this is, the event source queue. That queue will be processed by the dispatcher and events will be processed in order using the event's datetime (https://basana.readthedocs.io/en/latest/basana.html#basana.Event.when).

Going back to your example, the event generated by callback1  will have a greater datetime than the event currently being processed, so it will be processed after callback2 gets called.

```
import asyncio

import basana as bs


dispatcher = bs.backtesting_dispatcher()
source_1 = bs.FifoQueueEventSource()
source_2 = bs.FifoQueueEventSource()


async def callback_1(event: bs.Event):
    print("callback_1", event.when)
    source_2.push(bs.Event(bs.local_now()))


async def callback_2(event: bs.Event):
    print("callback_2", event.when)


async def callback_3(event: bs.Event):
    print("callback_3", event.when)


async def main():
    source_1.push(bs.Event(bs.local_now()))

    dispatcher.subscribe(source_1, callback_1)
    dispatcher.subscribe(source_1, callback_2)
    dispatcher.subscribe(source_2, callback_3)
    await dispatcher.run()


if __name__ == "__main__":
    asyncio.run(main())
```

```
callback_1 2024-12-21 13:52:52.947639-03:00
callback_2 2024-12-21 13:52:52.947639-03:00
callback_3 2024-12-21 13:52:52.948041-03:00
```

Let me know if this helps.

**Reply by cmzxo** [reply]
Very helpful! I should write this test case by myself, thanks for your time!
It's clear for me now that events are processed based on their datetime,
will play more with this library this week.
